【バグ】 with Apple's clang 15.0.0, use of an incomplete type, BaseItemDefinition, in baseitem-list.cpp

## 環境

- OS: macOS 14.7.2
- バージョン: 3.0.1Beta23 + changes up to 4d85bb6bddf40f48341f83be7c9e37eded1235fc + changes for macOS from https://github.com/backwardsEric/hengband ; there are no macOS-specific changes to util/abstract-vector-wrapper.h, system/baseitem/baseitem-list.h, or system/baseitem/baseitem-list.cpp

## バグの詳細

When compiling on Github's macos-latest runner (macOS 14.7.2, Xcode 15.4, Apple's clang 15.0.0), compiling fails in system/baseitem/baseitem-list.cpp when it includes either stdafx.h (with precompiled headers) or system/baseitem/baseitem-list.h (without precompiled headers).  The errors are all due to an incomplete type for BaseItemDefinition.

## 再現手順

Push a change that includes this workflow to the branch named 'baseitem-compiling-error' on a GitHub repository that has actions enabled:

```
name: compiler-error

on:
  push:
    branches: [ 'baseitem-compiling-error' ]

jobs:
  pch:
    runs-on: macos-latest

    steps:
      - name: Build
        run: |
          brew install m4
          brew install autoconf
          brew install automake
          brew install nkf
          git clone https://github.com/backwardsEric/hengband.git hengband
          cd hengband
          git submodule init
          git submodule update
          ./bootstrap
          ./configure CXXFLAGS="-g -O2 -Wno-invalid-source-encoding"
          make

  no-pch:
    runs-on: macos-latest

    steps:
      - name: Build
        run: |
          brew install m4
          brew install autoconf
          brew install automake
          brew install nkf
          git clone https://github.com/backwardsEric/hengband.git hengband
          cd hengband
          git submodule init
          git submodule update
          ./bootstrap
          ./configure CXXFLAGS="-g -O2 -Wno-invalid-source-encoding" --disable-pch
          make
```

For me, the errors that generates for the run with precompiled headers are:

```
In file included from system/baseitem/baseitem-list.cpp:1:
In file included from /Users/runner/work/hengband/hengband/hengband/src/stdafx.cpp:1:
In file included from /Users/runner/work/hengband/hengband/hengband/src/stdafx.h:5:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/functional:526:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:27:
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:944:62: error: arithmetic on a pointer to an incomplete type 'BaseitemDefinition'
        __alloc_traits::destroy(__alloc(), std::__to_address(--__soon_to_be_end));
                                                             ^ ~~~~~~~~~~~~~~~~
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:938:29: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__base_destruct_at_end' requested here
  void __clear() _NOEXCEPT {__base_destruct_at_end(this->__begin_);}
                            ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:489:20: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__clear' requested here
            __vec_.__clear();
                   ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:500:67: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__destroy_vector::operator()' requested here
  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI ~vector() { __destroy_vector(*this)(); }
                                                                  ^
../src/system/baseitem/baseitem-list.h:24:5: note: in instantiation of member function 'std::vector<BaseitemDefinition>::~vector' requested here
    ~BaseitemList() = default;
    ^
../src/system/baseitem/baseitem-list.h:24:5: note: in defaulted destructor for 'BaseitemList' first required here
../src/system/baseitem/baseitem-list.h:16:7: note: forward declaration of 'BaseitemDefinition'
class BaseitemDefinition;
      ^
In file included from system/baseitem/baseitem-list.cpp:1:
In file included from /Users/runner/work/hengband/hengband/hengband/src/stdafx.cpp:1:
In file included from /Users/runner/work/hengband/hengband/hengband/src/stdafx.h:5:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/functional:526:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:27:
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:607:52: error: arithmetic on a pointer to an incomplete type 'BaseitemDefinition'
        {return static_cast<size_type>(__end_cap() - this->__begin_);}
                                       ~~~~~~~~~~~ ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:871:56: note: in instantiation of member function 'std::vector<BaseitemDefinition>::capacity' requested here
      __annotate_contiguous_container(data(), data() + capacity(),
                                                       ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:490:20: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__annotate_delete' requested here
            __vec_.__annotate_delete();
                   ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:500:67: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__destroy_vector::operator()' requested here
  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI ~vector() { __destroy_vector(*this)(); }
                                                                  ^
../src/system/baseitem/baseitem-list.h:24:5: note: in instantiation of member function 'std::vector<BaseitemDefinition>::~vector' requested here
    ~BaseitemList() = default;
    ^
../src/system/baseitem/baseitem-list.h:24:5: note: in defaulted destructor for 'BaseitemList' first required here
../src/system/baseitem/baseitem-list.h:16:7: note: forward declaration of 'BaseitemDefinition'
class BaseitemDefinition;
      ^
In file included from system/baseitem/baseitem-list.cpp:1:
In file included from /Users/runner/work/hengband/hengband/hengband/src/stdafx.cpp:1:
In file included from /Users/runner/work/hengband/hengband/hengband/src/stdafx.h:5:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/functional:526:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:27:
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:871:54: error: arithmetic on a pointer to an incomplete type 'const value_type' (aka 'const BaseitemDefinition')
      __annotate_contiguous_container(data(), data() + capacity(),
                                              ~~~~~~ ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:490:20: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__annotate_delete' requested here
            __vec_.__annotate_delete();
                   ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:500:67: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__destroy_vector::operator()' requested here
  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI ~vector() { __destroy_vector(*this)(); }
                                                                  ^
../src/system/baseitem/baseitem-list.h:24:5: note: in instantiation of member function 'std::vector<BaseitemDefinition>::~vector' requested here
    ~BaseitemList() = default;
    ^
../src/system/baseitem/baseitem-list.h:24:5: note: in defaulted destructor for 'BaseitemList' first required here
../src/system/baseitem/baseitem-list.h:16:7: note: forward declaration of 'BaseitemDefinition'
class BaseitemDefinition;
      ^
In file included from system/baseitem/baseitem-list.cpp:1:
In file included from /Users/runner/work/hengband/hengband/hengband/src/stdafx.cpp:1:
In file included from /Users/runner/work/hengband/hengband/hengband/src/stdafx.h:2:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/algorithm:1811:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_any_all_none_of.h:12:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_find.h:14:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_backend.h:12:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_backends/cpu_backend.h:49:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_backends/cpu_backends/any_of.h:14:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_backends/cpu_backends/backend.h:20:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_backends/cpu_backends/libdispatch.h:20:
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__memory/allocator.h:130:58: error: invalid application of 'sizeof' to an incomplete type 'BaseitemDefinition'
            _VSTD::__libcpp_deallocate((void*)__p, __n * sizeof(_Tp), _LIBCPP_ALIGNOF(_Tp));
                                                         ^~~~~~~~~~~
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__memory/allocator_traits.h:288:13: note: in instantiation of member function 'std::allocator<BaseitemDefinition>::deallocate' requested here
        __a.deallocate(__p, __n);
            ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:491:29: note: in instantiation of member function 'std::allocator_traits<std::allocator<BaseitemDefinition>>::deallocate' requested here
            __alloc_traits::deallocate(__vec_.__alloc(), __vec_.__begin_, __vec_.capacity());
                            ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:500:67: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__destroy_vector::operator()' requested here
  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI ~vector() { __destroy_vector(*this)(); }
                                                                  ^
../src/system/baseitem/baseitem-list.h:24:5: note: in instantiation of member function 'std::vector<BaseitemDefinition>::~vector' requested here
    ~BaseitemList() = default;
    ^
../src/system/baseitem/baseitem-list.h:24:5: note: in defaulted destructor for 'BaseitemList' first required here
../src/system/baseitem/baseitem-list.h:16:7: note: forward declaration of 'BaseitemDefinition'
class BaseitemDefinition;
      ^
```

## 期待される正常な動作

baseitem-list.cpp compiles successfully.

## 補足事項

Compilation of the same source code does work with Xcode 14.2 and Apple clang 14.0.0 on macOS 12.7.6.

With that line deleted from baseitem-list.h, there's still compilation errors.  They appear too be the same:

```
In file included from system/baseitem/baseitem-list.cpp:1:
In file included from /Users/runner/work/hengband/hengband/src/stdafx.cpp:1:
In file included from /Users/runner/work/hengband/hengband/src/stdafx.h:5:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/functional:526:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:27:
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:944:62: error: arithmetic on a pointer to an incomplete type 'BaseitemDefinition'
        __alloc_traits::destroy(__alloc(), std::__to_address(--__soon_to_be_end));
                                                             ^ ~~~~~~~~~~~~~~~~
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:938:29: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__base_destruct_at_end' requested here
  void __clear() _NOEXCEPT {__base_destruct_at_end(this->__begin_);}
                            ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:489:20: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__clear' requested here
            __vec_.__clear();
                   ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:500:67: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__destroy_vector::operator()' requested here
  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI ~vector() { __destroy_vector(*this)(); }
                                                                  ^
./system/baseitem/baseitem-list.h:18:7: note: in instantiation of member function 'std::vector<BaseitemDefinition>::~vector' requested here
class BaseitemList : public util::AbstractVectorWrapper<BaseitemDefinition> {
      ^
./system/baseitem/baseitem-list.h:18:7: note: in implicit destructor for 'BaseitemList' first required here
./system/baseitem/baseitem-list.h:16:7: note: forward declaration of 'BaseitemDefinition'
class BaseitemDefinition;
      ^
In file included from system/baseitem/baseitem-list.cpp:1:
In file included from /Users/runner/work/hengband/hengband/src/stdafx.cpp:1:
In file included from /Users/runner/work/hengband/hengband/src/stdafx.h:5:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/functional:526:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:27:
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:607:52: error: arithmetic on a pointer to an incomplete type 'BaseitemDefinition'
        {return static_cast<size_type>(__end_cap() - this->__begin_);}
                                       ~~~~~~~~~~~ ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:871:56: note: in instantiation of member function 'std::vector<BaseitemDefinition>::capacity' requested here
      __annotate_contiguous_container(data(), data() + capacity(),
                                                       ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:490:20: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__annotate_delete' requested here
            __vec_.__annotate_delete();
                   ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:500:67: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__destroy_vector::operator()' requested here
  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI ~vector() { __destroy_vector(*this)(); }
                                                                  ^
./system/baseitem/baseitem-list.h:18:7: note: in instantiation of member function 'std::vector<BaseitemDefinition>::~vector' requested here
class BaseitemList : public util::AbstractVectorWrapper<BaseitemDefinition> {
      ^
./system/baseitem/baseitem-list.h:18:7: note: in implicit destructor for 'BaseitemList' first required here
./system/baseitem/baseitem-list.h:16:7: note: forward declaration of 'BaseitemDefinition'
class BaseitemDefinition;
      ^
In file included from system/baseitem/baseitem-list.cpp:1:
In file included from /Users/runner/work/hengband/hengband/src/stdafx.cpp:1:
In file included from /Users/runner/work/hengband/hengband/src/stdafx.h:5:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/functional:526:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:27:
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:871:54: error: arithmetic on a pointer to an incomplete type 'const value_type' (aka 'const BaseitemDefinition')
      __annotate_contiguous_container(data(), data() + capacity(),
                                              ~~~~~~ ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:490:20: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__annotate_delete' requested here
            __vec_.__annotate_delete();
                   ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:500:67: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__destroy_vector::operator()' requested here
  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI ~vector() { __destroy_vector(*this)(); }
                                                                  ^
./system/baseitem/baseitem-list.h:18:7: note: in instantiation of member function 'std::vector<BaseitemDefinition>::~vector' requested here
class BaseitemList : public util::AbstractVectorWrapper<BaseitemDefinition> {
      ^
./system/baseitem/baseitem-list.h:18:7: note: in implicit destructor for 'BaseitemList' first required here
./system/baseitem/baseitem-list.h:16:7: note: forward declaration of 'BaseitemDefinition'
class BaseitemDefinition;
      ^
In file included from system/baseitem/baseitem-list.cpp:1:
In file included from /Users/runner/work/hengband/hengband/src/stdafx.cpp:1:
In file included from /Users/runner/work/hengband/hengband/src/stdafx.h:2:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/algorithm:1811:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_any_all_none_of.h:12:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_find.h:14:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_backend.h:12:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_backends/cpu_backend.h:49:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_backends/cpu_backends/any_of.h:14:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_backends/cpu_backends/backend.h:20:
In file included from /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__algorithm/pstl_backends/cpu_backends/libdispatch.h:20:
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__memory/allocator.h:130:58: error: invalid application of 'sizeof' to an incomplete type 'BaseitemDefinition'
            _VSTD::__libcpp_deallocate((void*)__p, __n * sizeof(_Tp), _LIBCPP_ALIGNOF(_Tp));
                                                         ^~~~~~~~~~~
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__memory/allocator_traits.h:288:13: note: in instantiation of member function 'std::allocator<BaseitemDefinition>::deallocate' requested here
        __a.deallocate(__p, __n);
            ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:491:29: note: in instantiation of member function 'std::allocator_traits<std::allocator<BaseitemDefinition>>::deallocate' requested here
            __alloc_traits::deallocate(__vec_.__alloc(), __vec_.__begin_, __vec_.capacity());
                            ^
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:500:67: note: in instantiation of member function 'std::vector<BaseitemDefinition>::__destroy_vector::operator()' requested here
  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI ~vector() { __destroy_vector(*this)(); }
                                                                  ^
./system/baseitem/baseitem-list.h:18:7: note: in instantiation of member function 'std::vector<BaseitemDefinition>::~vector' requested here
class BaseitemList : public util::AbstractVectorWrapper<BaseitemDefinition> {
      ^
./system/baseitem/baseitem-list.h:18:7: note: in implicit destructor for 'BaseitemList' first required here
./system/baseitem/baseitem-list.h:16:7: note: forward declaration of 'BaseitemDefinition'
class BaseitemDefinition;
      ^
```

As might be expected, including system/baseitem/baseitem-definition.h before system/baseitem/baseitem-list.h in system/baseitem/baseitem-list.cpp does avoid the errors.

--- a/src/cmd-action/cmd-mane.cpp
+++ b/src/cmd-action/cmd-mane.cpp
@@ -1216,15 +1216,12 @@ static bool use_mane(PlayerType *player_ptr, MonsterAbilityType spell)
         break;
     }
     case MonsterAbilityType::S_DEAD_UNIQUE: {
-        auto count = 0;
         if (!target_set(player_ptr, TARGET_KILL)) {
             return false;
         }
         msg_print(_("特別な強敵を蘇生した！", "You summon special dead opponents!"));
         for (auto k = 0; k < 4; k++) {
-            if (summon_specific(player_ptr, target_row, target_col, plev, SUMMON_DEAD_UNIQUE, (mode | PM_ALLOW_UNIQUE | PM_CLONE))) {
-                count++;
-            }
+            summon_specific(player_ptr, target_row, target_col, plev, SUMMON_DEAD_UNIQUE, (mode | PM_ALLOW_UNIQUE | PM_CLONE));
         }
         break;
     }

--- a/src/system/baseitem/baseitem-list.cpp
+++ b/src/system/baseitem/baseitem-list.cpp
@@ -25,6 +25,8 @@ constexpr auto INVALID_BASEITEM_KEY = "Invalid Baseitem Key is specified! Type: 
 
 BaseitemList BaseitemList::instance{};
 
+BaseitemList::~BaseitemList() = default;
+
 BaseitemList &BaseitemList::get_instance()
 {
     return instance;

--- a/src/system/baseitem/baseitem-list.h
+++ b/src/system/baseitem/baseitem-list.h
@@ -39,7 +39,7 @@ private:
     BaseitemList() = default;
 
     static BaseitemList instance;
-    std::vector<BaseitemDefinition> baseitems{};
+    std::vector<BaseitemDefinition> baseitems;
 
     std::vector<BaseitemDefinition> &get_inner_container() override
     {

--- a/src/system/baseitem/baseitem-list.h
+++ b/src/system/baseitem/baseitem-list.h
@@ -21,7 +21,7 @@ public:
     BaseitemList(const BaseitemList &) = delete;
     BaseitemList &operator=(const BaseitemList &) = delete;
     BaseitemList &operator=(BaseitemList &&) = delete;
-    ~BaseitemList() = default;
+    ~BaseitemList();
 
     static BaseitemList &get_instance();
     BaseitemDefinition &get_baseitem(const short bi_id);

--- a/src/system/terrain/terrain-list.h
+++ b/src/system/terrain/terrain-list.h
@@ -36,7 +36,7 @@ private:
     TerrainList() = default;
 
     static TerrainList instance;
-    std::vector<TerrainType> terrains{};
+    std::vector<TerrainType> terrains;
     std::map<TerrainTag, short> tags; //!< @details 全てのTerrainTag を繰り込んだら、terrains からlookupが可能になる. そうなったら削除する.
 
     std::vector<TerrainType> &get_inner_container() override

